pkg/traceql/combine.go

```diff
@@ -48,6 +102,130 @@ func (c *MetadataCombiner) Metadata() []*tempopb.TraceSearchMetadata {
 	return m
 }
 
+// MetadataAfter returns all traces that started after the given time. anyCombiner has no concept of time so it just returns all traces
+func (c *anyCombiner) MetadataAfter(_ uint32) []*tempopb.TraceSearchMetadata {
+	return c.Metadata()
+}
+
+type mostRecentCombiner struct {
+	trs            map[string]*tempopb.TraceSearchMetadata
+	trsSorted      []*tempopb.TraceSearchMetadata
+	keepMostRecent int
+}
+
+func newMostRecentCombiner(limit int) *mostRecentCombiner {
+	return &mostRecentCombiner{
+		trs:            make(map[string]*tempopb.TraceSearchMetadata, limit),
+		trsSorted:      make([]*tempopb.TraceSearchMetadata, 0, limit),
+		keepMostRecent: limit,
+	}
+}
+
+// addSpanset adds a new spanset to the combiner. It only performs the asTraceSearchMetadata
+// conversion if the spanset will be added
+func (c *mostRecentCombiner) addSpanset(new *Spanset) {
+	// if we're not configured to keep most recent then just add it
+	if c.keepMostRecent == 0 || c.Count() < c.keepMostRecent {
+		c.AddMetadata(asTraceSearchMetadata(new))
+		return
+	}
+
+	// else let's see if it's worth converting this to a metadata and adding it
+	// if it's already in the list, then we should add it
+	if _, ok := c.trs[util.TraceIDToHexString(new.TraceID)]; ok {
+		c.AddMetadata(asTraceSearchMetadata(new))
+		return
+	}
+
+	// if it's within range
+	if c.OldestTimestampNanos() <= new.StartTimeUnixNanos {
+		c.AddMetadata(asTraceSearchMetadata(new))
+		return
+	}
+
+	// this spanset is too old to bother converting and adding it
+}
+
+// AddMetadata adds the new metadata to the map. if it already exists
+// use CombineSearchResults to combine the two
+func (c *mostRecentCombiner) AddMetadata(new *tempopb.TraceSearchMetadata) bool {
+	if existing, ok := c.trs[new.TraceID]; ok {
+		combineSearchResults(existing, new)
+		return true
+	}
+
+	if c.Count() == c.keepMostRecent && c.keepMostRecent > 0 {
+		// if this is older than the oldest element, bail
+		if c.OldestTimestampNanos() > new.StartTimeUnixNano {
```

Comment:
yes, these are good callouts. i wrote this logic thinking about the combiner being used in the engine, but the truth is there is all kinds of weird behavior with fractured traces when used in the query frontend. i don't know if there are good answers for all these edge cases without reconsiderations at the storage layer. this may be a feature that is _always_ best effort.

This is weird:
- Add metadata for trace id 1234 to the most recent list
- It gets pushed out by more recent traces
- Receive a new shard for 1234 that re-qualifies it for the most recent list
- We correctly return 1234 but have lost some information about it

Or this:
- Receive 10 "most recent" results and return a result to the client
- There is a trace shard earlier in the search window which pushes one of the returned results out of the most recent list.
- We have incorrectly returned a trace that should not have been in the list.